tests/test-exoskeleton-add-rule.php

```diff
@@ -0,0 +1,279 @@
+	 */
+	static function setUpBeforeClass() {
+		parent::setUpBeforeClass();
+
+		self::$three_valid_rules = array(
+			[
+				'route' => '/wp/v2/posts',
+				'window' => 10,
+				'limit'	=> 25,
+				'lockout' => 200,
+				'method' => 'any',
+			],[
+				'route' => '/wp/v2/post/1',
+				'window' => 100,
+				'limit'	=> 5,
+				'lockout' => 60,
+				'method' => 'GET',
+			],[
+				'route' => '/wp/v2/post/2',
+				'window' => 90,
+				'limit'	=> 2,
+				'lockout' => 30,
+				'method' => 'GET',
+			],
+		);
+	}
+
+	/**
+	 * Pre test setup
+	 */
+	function setUp() {
+		$instance = Exoskeleton::get_instance();
+		$instance->rules = [];
+
+	}
+
+	/**
+	 * Get instance of Exoskeleton
+	 */
+	function test_get_instance() {
+		$this->assertInstanceOf( Exoskeleton::class, Exoskeleton::get_instance() );
+	}
+
+	/**
+	 * Test adding a valid rule
+	 *
+	 * @dataProvider validRuleProvider
+	 * @param array $rule Valid exoskeleton rule definition.
+	 */
+	function test_add_valid_rule( $rule ) {
+
+		$this->assertTrue( exoskeleton_add_rule( $rule ) );
+	}
+
+	/**
+	 * Test adding a valid rule with multiple methods
+	 *
+	 * @dataProvider validRuleProvider
+	 * @param array $rule Valid exoskeleton rule definition.
+	 */
+	function test_add_valid_rule_with_multiple_methods( $rule ) {
+		$rule['method'] = 'POST,GET';
+		$this->assertTrue( exoskeleton_add_rule( $rule ) );
+	}
+
+	/**
+	 * Test adding rule fails if any required field is missing
+	 *
+	 * @dataProvider requiredFieldsProvider
+	 * @param string $field required field names.
+	 */
+	public function test_adding_rule_fails_missing_required_fields( $field ) {
+		$rule = $this->getValidRuleHelper();
+		unset( $rule[ $field ] );
+		$this->assertFalse( exoskeleton_add_rule( $rule ) );
+	}
+
+	/**
+	 * Test adding multiple valid rules
+	 */
+	function test_adding_multiple_valid_rules() {
+		$ruleset = $this->validRuleProvider();
+		$this->assertNull( exoskeleton_add_rules( $ruleset[0] ) );
+		$instance = Exoskeleton::get_instance();
+		$this->assertEquals( 3, count( $instance->rules ) );
+	}
+
+
+	/**
+	 * Test adding multiple rules including invalid
+	 */
+	function test_adding_multiple_rules_including_invalid() {
+		$ruleset = $this->validRuleProvider();
+		unset( $ruleset[0][1]['lockout'] );
+		$this->assertNull( exoskeleton_add_rules( $ruleset[0] ) );
+		$instance = Exoskeleton::get_instance();
+		$this->assertEquals( 2, count( $instance->rules ) );
+	}
+
+	/**
+	 * Test adding a valid rule for a custom route
+	 * Exoskeleton makes no check for existence of custom route.
+	 *
+	 * @dataProvider validRuleProvider
+	 * @param array $rule Valid exoskeleton rule definition.
+	 */
+	function test_add_valid_custom_route_rule( $rule ) {
+		$rule['route'] = '/custom/route/that/does/not/exist';
+
+		$this->assertTrue( exoskeleton_add_rule( $rule ) );
+	}
+
+	/**
+	 * Check that exoskeleton will not overwrite an already existing rule
+	 *
+	 * @dataProvider validRuleProvider
```

Comment:
DataProvider usage here doesn't provide additional coverage we could remove this  as additional executions on different rules don't provide benefit.